shared/lib_iec61853.cpp

```diff
@@ -858,27 +858,32 @@ bool iec61853_module_t::operator() ( pvinput_t &input, double TcellC, double opv
 	/* initialize output first */
 	out.Power = out.Voltage = out.Current = out.Efficiency = out.Voc_oper = out.Isc_oper = 0.0;
 	
-	double poa, tpoa, iamf;
-	iamf = 1;
+	double iamf_beam = 1;
+	double iamf_diff = 1;
+	double iamf_gnd = 1;
 
-	if( input.radmode != 3 ){ // Skip module cover effects if using POA reference cell data 
-		// plane of array irradiance, W/m2
-		poa = input.Ibeam + input.Idiff + input.Ignd; 
-
-		// transmitted poa through module cover
-		tpoa = poa;
+	double AOIModifier = 1;
 
-		if ( input.IncAng > AOI_MIN && input.IncAng < AOI_MAX )
-		{
-			iamf = iam( input.IncAng, GlassAR );
-			tpoa = poa - ( 1.0 - iamf )*input.Ibeam*cos(input.IncAng*3.1415926/180.0);
-			if( tpoa < 0.0 ) tpoa = 0.0;
-			if( tpoa > poa ) tpoa = poa;
-		}
+	double poa, tpoa;
+	if( input.radmode != 3 ){ // Skip module cover effects if using POA reference cell data
+		// plane of array irradiance, W/m2
+		poa = input.Ibeam + input.Idiff + input.Ignd;
+
+		iamf_beam = iam( input.IncAng, GlassAR );
+		// Add consideration of sky diffuse and ground components to match both the cec and mlm module models - CZ 2019
+		double theta_diff = (59.7 - 0.1388 * input.Tilt + 0.001497 * pow(input.Tilt, 2)); // from [2], equation 5.4.2
+	    double theta_gnd = (90.0 - 0.5788 * input.Tilt + 0.002693 * pow(input.Tilt, 2)); // from [2], equation 5.4.1
+		iamf_diff = iam( theta_diff, GlassAR );
+		iamf_gnd = iam( theta_gnd, GlassAR );
+
+        tpoa = iamf_beam * input.Ibeam + iamf_diff * input.Idiff + iamf_gnd * input.Ignd;
+        if( tpoa < 0.0 ) tpoa = 0.0;
+        if( tpoa > poa ) tpoa = poa;
 	
 		// spectral effect via AM modifier
 		double ama = air_mass_modifier( input.Zenith, input.Elev, AMA );
 		tpoa *= ama;
+		AOIModifier = tpoa/poa;
```

Comment:
This looks overall consistent with other models. From a code maintenance perspective, would it make sense to call calculateIrradianceThroughCoverDeSoto in lib_pv_incidence_modifier.cpp instead of re-implementing these equations? I didn't review in enough detail to see if there are any differences.

Reply:
@janinefreeman the calculateIrradianceThroughCoverDeSoto function does not consider the effects of antireflective coatings. Since most installed modules have them these days, it's important to capture those effects. It does seem worth it to modify the calculateIrradianceThroughCoverDeSoto function to account for ARC and then use this function in the iec61853, pvwatts and cec module models.